cpp/mg_utility/mage.hpp

```diff
@@ -0,0 +1,2757 @@
+
+  /// Construct Id from uint64_t
+  static Id FromUint(uint64_t id) { return Id(util::MemcpyCast<int64_t>(id)); }
+  /// Construct Id from int64_t
+  static Id FromInt(int64_t id) { return Id(id); }
+
+  int64_t AsInt() const { return id_; }
+  uint64_t AsUint() const { return util::MemcpyCast<uint64_t>(id_); }
+
+  bool operator==(const Id &other) const { return id_ == other.id_; }
+  bool operator!=(const Id &other) const { return !(*this == other); }
+
+  bool operator<(const Id &other) const { return id_ < other.id_; }
+
+ private:
+  explicit Id(int64_t id) : id_(id) {}
+
+  int64_t id_;
+};
+
+/// @brief Wrapper class for @ref mgp_graph.
+class Graph {
+ private:
+  friend class Node;
+  friend class Relationship;
+
+ public:
+  explicit Graph(mgp_graph *graph) : graph_(graph) {}
+
+  /// @brief Returns the graph order (number of nodes).
+  int64_t order() const;
+  /// @brief Returns the graph size (number of relationships).
+  int64_t size() const;
+
+  /// @brief Returns an iterable structure of the graph’s nodes.
+  GraphNodes nodes() const;
+  /// @brief Returns an iterable structure of the graph’s relationships.
+  GraphRelationships relationships() const;
+
+  /// @brief Returns the graph node with the given ID.
+  Node GetNodeById(const Id node_id) const;
+
+  /// @brief Returns whether the graph contains a node with the given ID.
+  bool Contains(const Id node_id) const;
+  /// @brief Returns whether the graph contains the given node.
+  bool Contains(const Node &node) const;
+  /// @brief Returns whether the graph contains the given relationship.
+  bool Contains(const Relationship &relationship) const;
+
+  /// @brief Returns whether the graph is mutable.
+  bool IsMutable() const { return mgp::graph_is_mutable(graph_); }
+  /// @brief Creates a node and adds it to the graph.
+  void CreateNode() { mgp::graph_create_vertex(graph_, memory); }
+  /// @brief Deletes a node from the graph.
+  void DeleteNode(const Node &node);
+  /// @brief Deletes a node and all its incident edges from the graph.
+  void DetachDeleteNode(const Node &node);
+  /// @brief Creates a relationship of type `type` between nodes `from` and `to` and adds it to the graph.
+  void CreateRelationship(const Node &from, const Node &to, const std::string_view type);
+  /// @brief Deletes a relationship from the graph.
+  void DeleteRelationship(const Relationship &relationship);
+
+ private:
+  mgp_graph *graph_;
+};
+
+/// @brief View of graph nodes; wrapper class for @ref mgp_vertices_iterator.
+class Nodes {
+ public:
+  explicit Nodes(mgp_vertices_iterator *nodes_iterator) : nodes_iterator_(nodes_iterator) {}
+
+  class Iterator {
+   public:
+    friend class Nodes;
+
+    explicit Iterator(mgp_vertices_iterator *nodes_iterator);
+    ~Iterator();
+    Iterator &operator++();
+    Iterator operator++(int);
+    bool operator==(Iterator other) const;
+    bool operator!=(Iterator other) const { return !(*this == other); }
+    Node operator*();
```

Comment:
Also, define traits above methods declarations, plus I think you need to return the reference to the underlying element since you might want it to be mutable from iterator point of view
```suggestion
    reference operator*();
```

Reply:
Also I think you should implement `operator->()` to support pointers as well

Reply:
With iterators, `operator->` returns the private pointer used by the iterator. 
In this case, that pointer is a MGP type `mgp_vertices_iterator *`, and returning it  would defeat the purpose of this API, which is abstracting MGP types away from the user. Therefore, I’ve left it unimplemented.

For the same reason, `operator*` returns `Node` instead of the reference type. `mgp_vertices_iterator *` contains MGP vertices which need to be converted into new API types.

I understand this is not conventional usage, but it is what had to be implemented to allow iteration with new API types.